<!-- 
Example launch file: Kinect Visual Odometry using fovis 
-->

<launch>
  
	<param name="/use_sim_time" value="true"/>
	<arg name="localpath" default="/home/edward/catkin_ws/src/pcl_registration/"/>
	
	
	<!-- Play Rosbag file -->
	<node pkg="rosbag" type="play" name="rosbag" args="-r 1 -s 15
		$(arg localpath)/3_level02_ccw/kinect_ship_test3_1.bag
		$(arg localpath)/3_level02_ccw/kinect_ship_test3_2.bag 
		$(arg localpath)/3_level02_ccw/kinect_ship_test3_3.bag  --clock "/>
	
	<node pkg="nodelet" type="nodelet" name="depth2cloud" args="standalone depth_image_proc/point_cloud_xyz">
	      <remap from="/camera_info" to="/camera/depth_registered/camera_info"/>
	      <remap from="/image_rect" to="/camera/depth_registered/image"/>
	      <remap from="points" to="/camera/depth/points2"/>
	</node>
	
	#############     For depthflow: X-forward, Z-downward    ####################
	############# Should convert to ICP coordinate as an initial guess ###########
	<node pkg="tf" type="static_transform_publisher" name="odom2camzforward"
	      args="0    0    0    1.57     0    1.57  /odom_frame /cam_zforward 100"/>   
	<node pkg="tf" type="static_transform_publisher" name="camzforward2caminit"
	      args="0    0    0    -1.57     -1.57    0  /cam_zforward /camera_init 100"/>
	<node pkg="tf" type="static_transform_publisher" name="camera2base" 
	      args="0    0    0    1.57  0    1.57  /camera /camera_base 100" />
	<node pkg="tf" type="static_transform_publisher" name="cambase2optical" 
	      args="0    0    0    0     0    0  /camera_base /camera_depth_optical_frame 100" />
	      
	
	############## For ICP: Z-forward, Y-downward ##################
	<node pkg="tf" type="static_transform_publisher" name="odom2asusinit"
	      args="0    0    0    1.57      0    1.57  /odom_frame /asus_init 100"/>
	<node pkg="tf" type="static_transform_publisher" name="asus2base"
	      args="0    0    0    -1.57   -1.57   0  /asus_base /base_frame 100" />

	<node pkg="tf" type="static_transform_publisher" name="base2optical"
	      args="0    0    0    1.57  0.0  1.57  /base_frame /camera_rgb_optical_frame 100" />


		
	######## Calculate the inter-frame motion estimation using depth flow #########
	<node pkg="ca_rangeflow" type="ca_rangeflow" name="ca_rangeflow" output="screen">
	      <rosparam file="$(find ca_rangeflow)/config/param640.yaml" command="load"/>
	      <remap from="/camera/depth/image_rect" to="/camera/depth_registered/image" />
	      <remap from="/camera/depth/camera_info" to="/camera/depth_registered/camera_info" />
	</node>

	
	############ Refine the motion estimation between frame-to-map using ICP ############
	<node pkg="ca_rangeflow" type="refine" name="refine" output="screen"  
	      args="input_cloud:=/camera/depth/points2">
	      <param name="odomFrame"           type="string"   value="asus_base" />
	      <param name="mapFrame"            type="string"   value="asus_init" />
	      
	      ############ Frame to Frame or ############
	      <param name="useMap"               type="bool"     value="true"/>
	      <param name="localMapNum"      	 type="int"      value="6000"/>
	      <param name="useRandom_sampling"   type="bool"     value="true"/>
	      <param name="random_sampling_num"  type="int"      value="3000"/>
	      <param name="voxel_leafsize"       type="double"   value="0.05"/>
	      
	      ############ ICP Parameters ##############
	      <param name="setMaxCorresDist"            type="double"   value="0.05"/>
	      <param name="rej_max_corresdist"          type="double"   value="0.03"/>
	      <param name="setMaxIterationNum"          type="int"      value="30"/>
	      <param name="setTransEpsilon"             type="double"   value="1e-8"/>
	      <param name="euclidean_fitness_epsilon"   type="double"   value="0.001"/>
	      
	      ############ local map resolution ##########
	      <param name="mapVoxelLeafSize"            type="double"   value="0.05"/>
	      <param name="skipNum"           		type="int" 	value="1"/>
	      <param name="refine"                      type="bool"     value="false" />
	      
	      ########### NDT Parameters ##################
	      <param name="useNDT"                   type="bool"        value="true" />
	      <param name="setNDTEpsilon"            type="double"      value="0.1"/>
	      <param name="setStepSize"              type="double" 	value="0.1"/>
	      <param name="setResolution"            type="double"      value="1.0" />
	      <param name="setMaxInteration"         type="int"         value="100" />
	        
	</node>	
	      

	<!--<node pkg="rviz" type="rviz" name="rviz" args="-d $(find ca_rangeflow)/launch/odom.rviz"/>-->



	<!--<node pkg="tf" type="static_transform_publisher" name="actualmap_to_cloudmap"-->
		  <!--args="0.0  0.0  0.0  3.14  0.0  -1.9 /map /asus_init  100"  />-->

	<!--<node pkg="tf" type="static_transform_publisher" name="actualmap_to_cloudmap"-->
		  <!--args="0.0  0.0  0.0  3.14  0.0  -1.87 /map /odom_frame  100"  />-->

	<node pkg="tf" type="static_transform_publisher" name="actualmap_to_cloudmap"
		  args="0.0  0.0  0.0  -3.14  -0.15  -1.87 /map /odom_frame 100"  />
	<!-- Show actual map in rviz-->
	<!--<node pkg="map_server" type="map_server" name="actual_map" args="$(find ca_rangeflow)/mapfiles/level2.yaml" />-->

	<node pkg="hector_trajectory_server" type="hector_trajectory_server" name="localization_path" output="screen">
		<param name="target_frame_name" type="string" value="map" />
		<param name="source_frame_name" type="string" value="camera_rgb_optical_frame" />
		<param name="trajectory_update_rate" type="double" value="10" />
		<param name="trajectory_publish_rate" type="double" value="10" />
	</node>

	<node pkg="rviz" type="rviz" name="rviz" args="-d $(find ca_rangeflow)/config/icp.rviz"/>
	<node pkg="map_server" type="map_server" name="actual_map" args="$(find ca_depth_odometry)/mapfiles/level2.yaml" />

	
</launch>
